test/staking-rewards/MerkleTreeRewards.ts

```diff
@@ -0,0 +1,222 @@
+    });
+
+    describe('rewards', () => {
+        let merkleTreeRewards: MerkleTreeRewards;
+
+        let provider1: SignerWithAddress;
+        let provider2: SignerWithAddress;
+        let provider3: SignerWithAddress;
+        let provider4: SignerWithAddress;
+        let provider5: SignerWithAddress;
+
+        let rewards: Record<string, BigNumber> = {};
+        let merkleTree: MerkleTree;
+
+        const generateLeaf = (address: string, amount: BigNumber) =>
+            Buffer.from(
+                solidityKeccak256(['address', 'uint256'], [getAddress(address), amount.toString()]).slice(2),
+                'hex'
+            );
+
+        before(async () => {
+            [provider1, provider2, provider3, provider4, provider5] = await ethers.getSigners();
+
+            rewards = {
+                [provider1.address]: BigNumber.from(1),
+                [provider2.address]: BigNumber.from(100_000),
+                [provider3.address]: BigNumber.from(500),
+                [provider4.address]: BigNumber.from(1_000_000),
+                [provider5.address]: BigNumber.from(5)
+            };
+
+            merkleTree = new MerkleTree(
+                Object.entries(rewards).map(([provider, amount]) => generateLeaf(provider, amount)),
+                keccak256,
+                { sortPairs: true }
+            );
+        });
+
+        beforeEach(async () => {
+            merkleTreeRewards = await Contracts.MerkleTreeRewards.deploy(
+                bancorNetworkV3.address,
+                bntGovernance.address,
+                merkleTree.getRoot()
+            );
+
+            await bntGovernance.grantRole(ROLE_MINTER, merkleTreeRewards.address);
+        });
+
+        const testClaimOrStake = (stake: boolean) => {
+            interface Overrides {
+                caller?: SignerWithAddress;
+                amount?: BigNumber;
+                proof?: string[];
+            }
+
+            const claim = async (provider: SignerWithAddress, overrides: Overrides = {}) => {
+                let { caller = provider, amount, proof } = overrides;
+
+                if (!amount) {
+                    amount = rewards[provider.address];
+                }
+
+                if (!proof) {
+                    proof = merkleTree.getHexProof(generateLeaf(provider.address, amount));
+                }
+
+                return stake
+                    ? merkleTreeRewards.connect(caller).stakeRewards(provider.address, amount, proof)
+                    : merkleTreeRewards.connect(caller).claimRewards(provider.address, amount, proof);
+            };
+
+            describe(stake ? 'staking' : 'claiming', () => {
+                it('should revert when calling for a different provider', async () => {
+                    await expect(claim(provider1, { caller: provider2 })).to.be.revertedWith('AccessDenied');
+                });
+
+                it('should revert when calling with a different amount', async () => {
```

Comment:
can we add a test with a different amount but a valid proof for that amount (but of course different proof than in the contract)?